### DISCUSSION TITLE
How to extract a stream object from a PDF file

### DISCUSSION THREAD

**Discussion by jcdutton** [qpdf/qpdf#648] [Q&A]
Hi,
I want to load a pdf file, and list the objects in that file, and then traverse the tree to find the "font" object, and load that into RAM. 
I am new to libqpdf, so any pointers or examples I could start with.
I wish to look in the "font"  objects, and from that create /ToUnicode objects.
Some PDF files, with subset fonts, are difficult to cut&paste, mainly because they lack /ToUnicode objects.
So, I wish to "repair" PDF files such that cut&paste works.
libqpdf seems a good basis to start from.
Also, does libqpdf already support the creation of /ToUnicode objects, or will I need to write some code for that?

**Comment by jberkenbilt** [comment]
You can definitely use libqpdf to find font objects and load them into RAM. You'll want to read through [examples](https://github.com/qpdf/qpdf/tree/main/examples) for ideas on how to get started. In your case, you'll probably want to start with `getAllPages()` (see [QPDF.hh](https://github.com/qpdf/qpdf/blob/2b8d0f385b56d2a7307679ace4c50adbdbbddd03/include/qpdf/QPDF.hh#L599)) or look at [QPDFPageDocumentHelper.hh](https://github.com/qpdf/qpdf/blob/main/include/qpdf/QPDFPageDocumentHelper.hh). Once you have the object, you use the various methods of QPDFObjectHandle to access the object. If you want to load a stream, you can use `getStreamData`, a `QPDFObjectHandle` method.

libqpdf does not have support for creation of `/ToUnicode` objects, so you'll have to write code for that, but once you have it, you can easily use libqpdf to add the object to the file and hook it into the font dictionary.

Hopefully this is enough to get you started.
